extensions/amp-ad-network-doubleclick-impl/0.1/safeframe-host.js

```diff
@@ -0,0 +1,580 @@
+import {dev} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getData} from '../../../src/event-helper';
+import {tryParseJson} from '../../../src/json';
+
+/**
+ * Used to manage messages for different Safeframe ad slots.
+ *
+ * Maps a sentinel value to an instance of the SafeframeHostApi to which that
+ * sentinel value belongs.
+ * @type{!Object<string,SafeframeHostApi>}
+ */
+const safeframeHosts = {};
+
+let safeframeListenerCreated = false;
+
+const MESSAGE_FIELDS = {
+  CHANNEL: 'c',
+  SENTINEL: 'e',
+  ENDPOINT_IDENTITY: 'i',
+  PAYLOAD: 'p',
+  SERVICE: 's',
+  MESSAGE: 'message',
+};
+
+const SERVICE = {
+  GEOMETRY_UPDATE: 'geometry_update',
+  CREATIVE_GEOMETRY_UPDATE: 'creative_geometry_update',
+  EXPAND_REQUEST: 'expand_request',
+  EXPAND_RESPONSE: 'expand_response',
+  REGISTER_DONE: 'register_done',
+  COLLAPSE_REQUEST: 'collapse_request',
+  COLLAPSE_RESPONSE: 'collapse_response',
+};
+
+const TAG = 'AMP-DOUBLECLICK-SAFEFRAME';
+
+/** @const {string} */
+export const SAFEFRAME_ORIGIN = 'https://tpc.googlesyndication.com';
+
+/**
+ * Event listener callback for message events. If message is a Safeframe message,
+ * handles the message.
+ * This listener is registered within SafeframeHostApi.
+ */
+function safeframeListener() {
+  const data = tryParseJson(getData(event));
+  /** Only process messages that are valid Safeframe messages */
+  if (event.origin != SAFEFRAME_ORIGIN || !data) {
+    return;
+  }
+  /**
+   * If the sentinel is provided at the top level, this is a message simply
+   * to setup the postMessage channel, so set it up.
+   */
+  if (data[MESSAGE_FIELDS.SENTINEL]) {
+    receiveSetupMessage(data);
+  } else {
+    receiveStandardMessage(data);
+  }
+}
+
+/**
+ * Handles all non-setup messages.
+ * @param {Object} data Message data from the postMessage received from
+ *   the safeframe.
+ */
+function receiveStandardMessage(data) {
+  const payload = tryParseJson(data[MESSAGE_FIELDS.PAYLOAD]);
+  if (!payload || !payload['sentinel']) {
+    return;
+  }
+  const safeframeHost = safeframeHosts[payload['sentinel']];
+  if (!safeframeHost) {
+    dev().warn(
+        TAG, `Safeframe Host for sentinel ${payload['sentinel']} not found.`);
+    return;
+  }
+  safeframeHost.processMessage(payload, data[MESSAGE_FIELDS.SERVICE]);
+}
+
+
+/**
+ * Handles the setup message from the safeframe.
+ * @param {Object} data Message data from the postMessage received from
+ *   the safeframe.
+ */
+function receiveSetupMessage(data) {
+  const safeframeHost = safeframeHosts[data[MESSAGE_FIELDS.SENTINEL]];
+  if (!safeframeHost) {
+    dev().warn(TAG, 'Safeframe Host for sentinel ' +
+               `${data[MESSAGE_FIELDS.SENTINEL]} not found.`);
+    return;
+  }
+  if (!safeframeHost.channel) {
+    safeframeHost.connectMessagingChannel(data[MESSAGE_FIELDS.CHANNEL]);
+  }
+}
+
+/**
+ * Sets up the host for GPT Safeframe.
```

Comment:
Suggest you add documentation on the lifecycle/message flow for safeframe initialization and geom/resize communication paths

Reply:
done